Add `type` and `allowSymlinks` options, only find files by default

coreyfarrell opened the pull request:
This is not ready to merge, just seeking early feedback.  The
conditional code for stat vs lstat is a bit much, can/should this
logic be shifted to path-type or another module?

coreyfarrell committed:
Add support for `type` and `followSymlinks` options

sindresorhus commented on test.js:
```diff
@@ -12,10 +13,38 @@ test('async', async t => {
+	t.is(await locatePath(['fixture'], {type: 'file'}), undefined);
+	t.is(await locatePath(['fixture']), 'fixture');
+
+	if (!isWindows()) {
```
No need for a dependency for this. Just do `if (process.platform !== 'win32') {}`

sindresorhus commented on index.js:
```diff
@@ -1,26 +1,68 @@
+	fifo: 'isFIFO',
+	file: 'isFile',
+	socket: 'isSocket',
+	symbolicLink: 'isSymbolicLink'
```
I can't really imagine anyone needing to use `locate-path` to find anything other than `file` or `directory`. This seems like overkill.

coreyfarrell committed:
Add TypeScript definitions and tests for new options

coreyfarrell commented:
@sindresorhus I've addressed your comments and modified readme.md, typescript definitions/tests.  I really don't know typescript so hopefully I got that right.

One final question I have: is `followSymlinks` the best name for this option?  The implementation uses this to control using `fs.stat` or `fs.lstat`, but does not cause the `fs.realpath` to be returned.  I think this is best for `locate-path` and `find-up`, returning `fs.realpath` would be an edge case breaking change.  Thinking about longer running processes that might use find-up to determine a path at startup then reopen at a later time.  If the symbolic link were pointed to a new destination I think they would want to get that updated destination.

coreyfarrell committed:
Fix checkType error handling, add test

sindresorhus commented on index.js:
```diff
@@ -1,26 +1,63 @@
+		try {
+			const stat = await statFn(path.resolve(options.cwd, path_));
+			return matchType(options.type, stat);
+		} catch (error) {
```
```suggestion
		} catch (_) {
```

sindresorhus commented on index.js:
```diff
@@ -1,26 +1,63 @@
+			if (matchType(options.type, stat)) {
+				return path_;
+			}
+		} catch (error) {
```
```suggestion
		} catch (_) {
```

sindresorhus commented on index.d.ts:
```diff
@@ -1,11 +1,29 @@
+
+		@default undefined
+		*/
+		readonly type?: OptionsType;
```
```suggestion
		readonly type?: 'file' | 'directory';
```

sindresorhus commented on index.d.ts:
```diff
@@ -1,11 +1,29 @@
+		/**
+		Type of file to match.
+
+		@default undefined
```
I hate to do another breaking change, but I think it's worth it for clarity, to default to `'file'`.

sindresorhus commented on index.d.ts:
```diff
@@ -1,11 +1,29 @@
 		readonly cwd?: string;
+
+		/**
+		Type of file to match.
```
This description is not correct. It's not just `file`.

I think we should also explain that it can be set to `undefined` to match either file or directory. However, is that really useful? Should we support that at all?

coreyfarrell replied:
I don't see any reason to support `undefined`, it was mostly done that way to avoid being a breaking change.

~~If this module will only match file or directory then I think `followSymlinks` is a bad name, `allowSymlinks` (default true) or `rejectSymlinks` (default false) would be better.~~

Looking at the description in `index.d.ts` again `followSymlinks` is probably fine.

sindresorhus replied:
Let’s go with `allowSymlinks`. And not support `undefined`.

coreyfarrell replied:
Done

sindresorhus commented on fixture/dir-link:
```diff
@@ -0,0 +1 @@
+.
```
`dir-link` => `directory-link`

coreyfarrell committed:
Address review

coreyfarrell committed:
Rename followSymlinks -> allowSymlinks

sindresorhus committed:
Update readme.md

sindresorhus committed:
Update test.js
